main/design_documents/function-multi-versioning.rst

```diff
@@ -0,0 +1,56 @@
+=======================================================
+Function Multi Versioning support for Arm architectures
+=======================================================
+
+I'd like to propose this addition to the ACLE. Other architectures supports
+function multi versioning[1]. In sort just FMV. All toolchains and platforms
+should use the same source code formats because without it the software
+development and deployment will be too complicated and the feature won't be
+used. Arm architectures are a bit more complex in view of feature availability
+therefore function versions would be better to be bound to features instead of
+cpu implementations. Also the kernels may limit the available features in the
+user space.
+
+Why ACLE instead of ABI? FMV shall not change how the function behaves from the
+ABI perspective. All versions shall have the same function signature, because
+the caller will not know which version is actually called.
+
+Today the ``target`` attribute is used for many purposes and once code will
+start use function multi versioning the not yet updated compilers may got
+confused by the new values therefore I propose the ``target_version`` besides
+the ``target_clones``.
+
+The ACLE won't specify the underlying mechanism of the selection and the
+dispatching. On an embedded system probably it will be implemented differently
+form a desktop operating system. On Linux based systems I expect it will use
+IFUNC[2]. While the ACLE should specify the behavior that software could rely on
+every implementation.
+
+Feature test macro
+##################
+
+In order for users to determine the availability of support in the toolchain,
+I would like to propose the addition of the following feature test macro, to be
+defined to 1 if the support is available:
+
+::
+
+  __ARM_FEATURE_FUNCTION_MULTI_VERSIONING
+
+I don't expect that enabling a feature for function then ACLE features will be
+available in the function scope if otherwise won't be enabled. This would require
+significant changes how the C/C++ preprocessor works.
+
+Attributes
+##########
+
+These use cases to be considered:
+- The compiler is not aware of the new attributes.
+Usually unknown attributes are ignored, but the feature test macro is there to
+disable the unwanted code parts.
+- FMV aware compiler processes the attributes with newer arch features.
+In this case the compiler should ignore the new feature because it won't able to
+process it.
```

Comment:
This doesn't render correctly in the html page: https://github.com/DanielKristofKiss/acle/blob/FMV/main/design_documents/function-multi-versioning.rst#attributes

Reply:
fixed.